apps/restconf/restconf_methods.c

```diff
@@ -581,6 +583,409 @@ api_data_write(clicon_handle h,
+  * @retval new cbuf containing the truncated string (old buffer remains as it was)
+  * @retval NULL   Error
+  */
+cbuf*
+cbuf_trunc(cbuf       *cb,
+		int       pos)
+{
+    if (pos < 0 || pos > cb->cb_strlen){
+	errno = EINVAL;
+	return NULL;
+    }
+    /* Ensure buffer is right size */
+    cbuf* new_buf = cbuf_new_alloc(pos + 1);
+    if (new_buf == NULL)
+        return NULL;
+    strncpy(new_buf->cb_buffer, cb->cb_buffer, pos);
+    new_buf->cb_strlen = pos;
+    return new_buf;
+}
+
+/*! Add square brackets after the surrounding curly brackets in JSON
+  * Needed, in order to modify the result of xml2json_cbuf() to be valid input
+  * to api_dta_post() and api_dta_write()
+  * @param [in]  x_simple_patch  a cxobj to pass to xml2json_cbuf()
+  * @retval new cbuf with the modified json
+  * @retval NULL   Error
+  */
+
+static cbuf* yang_patch_xml2json_modified_cbuf(cxobj* x_simple_patch)
+{
+    cbuf *json_simple_patch = cbuf_new();
+    if (json_simple_patch == NULL)
+        return NULL;
+    cbuf* cb = cbuf_new();
+    xml2json_cbuf(cb, x_simple_patch, 1);
+
+    // Insert a '[' after the first '{' to get the JSON to match what api_data_post/write() expect
+    char *json_simple_patch_tmp = cbuf_get(cb);
+    int brace_count = 0;
+    for (int l = 0; l < strlen(json_simple_patch_tmp); l++) {
+        char c = json_simple_patch_tmp[l];
+        if (c == '{') {
+            brace_count++;
+            if (brace_count == 2) {
+                cbuf_append(json_simple_patch,(int)'[');
+            }
+        }
+        cbuf_append(json_simple_patch,(int)c);
+    }
+    cbuf* json_simple_patch_2 = NULL;
+
+    // Insert a ']' before the last '}' to get the JSON to match what api_data_post() expects
+    for (int l = cbuf_len(json_simple_patch); l>= 0; l--) {
+        char c = cbuf_get(json_simple_patch)[l];
+        if (c == '}') {
+            // Truncate and add a string, as there is not a function to insert a char into a cbuf
+           json_simple_patch_2 = cbuf_trunc(json_simple_patch, l);
+           cbuf_append_str(json_simple_patch_2, "]}");
+           break;
+        }
+    }
+    cbuf_free(json_simple_patch);
+    cbuf_free(cb);
+    return json_simple_patch_2;
+}
+
+/*!yang_patch_strip_after_last_slash 
+ *
+ * Strip /... from end  of val
+ * so that e.g. "/interface=eth2" becomes "/"
+ * or "/interface_list=mylist/interface=eth2" becomes "/interface_list=mylist/"
+ *
+ * @param[in]  val         value to strip
+ * @retval new cbuf with the stripped string
+ * @retval NULL error
+ */
+static cbuf* yang_patch_strip_after_last_slash(cbuf* val)
+{
+      cbuf *cb = cbuf_new();
+      cbuf* val_tmp = cbuf_new();
+      cbuf_append_str(val_tmp, cbuf_get(val));
+      int idx = cbuf_len(val_tmp);
+      for (int l = cbuf_len(val_tmp); l>= 0; l--) {
+          if (cbuf_get(val_tmp)[l] == '/') {
+              idx = l;
+              break;
+          }
+      }
+      cbuf* val_tmp_2 = cbuf_trunc(val_tmp, idx + 1);
```

Comment:
idx already equal to cbuf_len, so idx + 1 is definitely out of range

Reply:
idx is not definitely out of range, since it gets a value in the loop after the initialization to cbuf_len. In the case that there is no slash, idx will not get a value in the loop, and will be out-of-range - which is a condition that I check in cbuf_trunc. However, I modified the code so that the out of range check is also done before calling cbuf_trunc().